arch/arm/core/cortex_m/mpu/arm_mpu_v7_internal.h

```diff
@@ -0,0 +1,225 @@
+ * the SIZE field value of MPU_RASR.
+ */
+static inline u32_t _size_to_mpu_rasr_size(u32_t size)
+{
+	/* The minimal supported region size is 32 bytes */
+	if (size <= 32) {
+		return REGION_32B;
+	}
+
+	/*
+	 * A size value greater than 2^31 could not be handled by
+	 * round_up_to_next_power_of_two() properly. We handle
+	 * it separately here.
+	 */
+	if (size > (1 << 31)) {
+		return REGION_4G;
+	}
+
+	size = 1 << (32 - __builtin_clz(size - 1));
+	return (32 - __builtin_clz(size) - 2) << 1;
+}
+
+/**
+ * Generate the value of the MPU Region Attribute and Size Register
+ * (MPU_RASR) that corresponds to the supplied MPU region attributes.
+ * This function is internal to the driver.
+ */
+static inline u32_t _get_region_attr(u32_t xn, u32_t ap, u32_t tex,
+				     u32_t c, u32_t b, u32_t s,
+				     u32_t srd, u32_t region_size)
+{
+	int size = _size_to_mpu_rasr_size(region_size);
+
+	return (((xn << MPU_RASR_XN_Pos) & MPU_RASR_XN_Msk)
+		| ((ap << MPU_RASR_AP_Pos) & MPU_RASR_AP_Msk)
+		| ((tex << MPU_RASR_TEX_Pos) & MPU_RASR_TEX_Msk)
+		| ((s << MPU_RASR_S_Pos) & MPU_RASR_S_Msk)
+		| ((c << MPU_RASR_C_Pos) & MPU_RASR_C_Msk)
+		| ((b << MPU_RASR_B_Pos) & MPU_RASR_B_Msk)
+		| ((srd << MPU_RASR_SRD_Pos) & MPU_RASR_SRD_Msk)
+		| (size));
+}
+
+/**
+ * This internal function allocates default RAM cache-ability, share-ability,
+ * and execution allowance attributes along with the requested access
+ * permissions and size.
+ */
+static inline void _get_mpu_ram_region_attr(arm_mpu_region_attr_t *p_attr,
+	u32_t ap, u32_t base, u32_t size)
+{
+	/* in ARMv7-M MPU the base address is not required
+	 * to determine region attributes.
+	 */
+	(void) base;
+
+	p_attr->rasr = _get_region_attr(1, ap, 1, 1, 1, 1, 0, size);
+}
+
+/**
+ * This internal function is utilized by the MPU driver to combine a given
+ * MPU attribute configuration and region size and return the correct
+ * parameter set.
+ */
+static inline void _get_region_attr_by_conf(arm_mpu_region_attr_t *p_attr,
+	u32_t attr, u32_t base, u32_t size)
+{
+	/* in ARMv7-M MPU the base address is not required
+	 * to determine region attributes
+	 */
+	(void) base;
+
+	p_attr->rasr = attr | _size_to_mpu_rasr_size(size);
+}
+
+/**
+ * This internal function checks if region is enabled or not.
+ *
+ * Note:
+ *   The caller must provide a valid region number.
+ */
+static inline int _is_enabled_region(u32_t r_index)
+{
+	MPU->RNR = r_index;
+
+	return MPU->RASR & MPU_RASR_ENABLE_Msk;
+}
+
+/**
+ * This internal function checks if the given buffer is in the region.
+ *
+ * Note:
+ *   The caller must provide a valid region number.
+ */
+static inline int _is_in_region(u32_t r_index, u32_t start, u32_t size)
+{
+	u32_t r_addr_start;
+	u32_t r_size_lshift;
+	u32_t r_addr_end;
+
+	MPU->RNR = r_index;
+	r_addr_start = MPU->RBAR & MPU_RBAR_ADDR_Msk;
+	r_size_lshift = ((MPU->RASR & MPU_RASR_SIZE_Msk) >>
+			MPU_RASR_SIZE_Pos) + 1;
+	r_addr_end = r_addr_start + (1 << r_size_lshift) - 1;
```

Comment:
1ul << ...

BTW: You can just compare base address with `(base + size - 1) & ~((1ul << shift) - 1)`.

Reply:
OK about 1UL.
I admit, the suggestion confuses me more :)